File: DBGenerator/DbContainer/DbTable.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;

namespace DBContainer
{
    public partial class DbTable
    {
        public string           name            = "";
        public string           comment         = "";
        public List<object>     columnDatas     = new List<object>();
        public List<DbColumn>   columns         = new List<DbColumn>();
        public List<List<bool>> nullflags       = new List<List<bool>>();
        public int              rowCount        = 0;

        public List<DbColumn> primaryKeys
        {
            get { return columns.Where( c => c.primaryKey ).ToList(); }
        }

        public DbTable AddColumn( DbColumn c )
        {
            columns.Add( c );
            nullflags.Add( new List<bool>() );

            switch( c.dataType )
            {
            case DbDataTypes.Boolean:
                columnDatas.Add( new List<bool>() ); break;
            case DbDataTypes.Int8:
                columnDatas.Add( new List<sbyte>() ); break;
            case DbDataTypes.Int16:
                columnDatas.Add( new List<short>() ); break;
            case DbDataTypes.Int32:
                columnDatas.Add( new List<int>() ); break;
            case DbDataTypes.Int64:
                columnDatas.Add( new List<long>() ); break;
            case DbDataTypes.UInt8:
                columnDatas.Add( new List<byte>() ); break;
            case DbDataTypes.UInt16:
                columnDatas.Add( new List<ushort>() ); break;
            case DbDataTypes.UInt32:
                columnDatas.Add( new List<uint>() ); break;
            case DbDataTypes.UInt64:
                columnDatas.Add( new List<ulong>() ); break;
            case DbDataTypes.Float:
                columnDatas.Add( new List<float>() ); break;
            case DbDataTypes.Double:
                columnDatas.Add( new List<double>() ); break;
            case DbDataTypes.DateTime:
                columnDatas.Add( new List<DateTime>() ); break;
            case DbDataTypes.String:
                columnDatas.Add( new List<string>() ); break;
            case DbDataTypes.Bytes:
                columnDatas.Add( new List<byte[]>() ); break;
            default:
                break;
            }

            return this;
        }

        public DbRow AddRow()
        {
            foreach( var col in columns )
            {
                switch( col.dataType )
                {
                case DbDataTypes.Boolean:
                    ( (List<bool>)columnDatas[ col.columnIndex ] ).Add( default( bool ) ); break;
                case DbDataTypes.Int8:
                    ( (List<sbyte>)columnDatas[ col.columnIndex ] ).Add( default( sbyte ) ); break;
                case DbDataTypes.Int16:
                    ( (List<short>)columnDatas[ col.columnIndex ] ).Add( default( short ) ); break;
                case DbDataTypes.Int32:
                    ( (List<int>)columnDatas[ col.columnIndex ] ).Add( default( int ) ); break;
                case DbDataTypes.Int64:
                    ( (List<long>)columnDatas[ col.columnIndex ] ).Add( default( long ) ); break;
                case DbDataTypes.UInt8:
                    ( (List<byte>)columnDatas[ col.columnIndex ] ).Add( default( byte ) ); break;
                case DbDataTypes.UInt16:
                    ( (List<ushort>)columnDatas[ col.columnIndex ] ).Add( default( ushort ) ); break;
                case DbDataTypes.UInt32:
                    ( (List<uint>)columnDatas[ col.columnIndex ] ).Add( default( uint ) ); break;
                case DbDataTypes.UInt64:
                    ( (List<ulong>)columnDatas[ col.columnIndex ] ).Add( default( ulong ) ); break;
                case DbDataTypes.Float:
                    ( (List<float>)columnDatas[ col.columnIndex ] ).Add( default( float ) ); break;
                case DbDataTypes.Double:
                    ( (List<double>)columnDatas[ col.columnIndex ] ).Add( default( double ) ); break;
                case DbDataTypes.DateTime:
                    ( (List<DateTime>)columnDatas[ col.columnIndex ] ).Add( default( DateTime ) ); break;
                case DbDataTypes.String:
                    ( (List<string>)columnDatas[ col.columnIndex ] ).Add( default( string ) ); break;
                case DbDataTypes.Bytes:
                    ( (List<byte[]>)columnDatas[ col.columnIndex ] ).Add( default( byte[] ) ); break;
                default:
                    break;
                }
                if( col.nullable )
                {
                    nullflags[ col.columnIndex ].Add( true );
                }
            }
            return new DbRow { parent = this, rowIndex = rowCount++ };
        }

        public DbRow this[ int rowIdx ]
        {
            get
            {
                if( rowIdx > rowCount ) throw new Exception( "out of range: rowIdx" );
                return new DbRow { parent = this, rowIndex = rowIdx };
            }
        }

        public int getIndexByColumnName( string name )
        {
            return columns.FindIndex( a => a.name == name );
        }

    }
}

File: Lib/DbTable.cpp
#include "All.h"

namespace xxx
{

    DbColumn& DbTable::AddColumn( DbDataTypes dataType )
    {
        auto& col = columns.Emplace();
        col.dataType = dataType;
        nullflags.Emplace();
        col.parent = this;
        col.columnIndex = columns.Size() - 1;
        columnDatas.Emplace( col.dataType );
        return col;
    }

    DbColumn& DbTable::AddColumn( String const& name, DbDataTypes dataType )
    {
        auto& col = AddColumn( dataType );
        col.name = name;
        return col;
    }

    void DbTable::AddRowCore( int& i, DbNull_t v )
    {
#if __DEBUG
        if( !columns[ i ].nullable )
        {
            throw std::exception( "column is not nullable" );
        }
#endif
        columnDatas[ i ].Grow();
        nullflags[ i ].Push( true );
    }

    DbRow DbTable::AddRow()
    {
        for( int i = 0; i < columns.Size(); ++i )
        {
            columnDatas[ i ].Grow();
            if( columns[ i ].nullable )
            {
                nullflags[ i ].Push( true );
            }
        }
        return DbRow{ this, rowCount++ };
    }


    void DbTable::Dump( String& output ) const
    {
        for( int j = 0; j < columns.Size(); ++j )
        {
            if( j ) output.Append( "\t" );
            if( columns[ j ].name.Size() )
            {
                output.Append( columns[ j ].name );
            }
            else
            {
                output.Append( "col_", j );
            }
        }
        output.Append( '\n' );
        for( int i = 0; i < rowCount; ++i )
        {
            if( i ) output.Append( '\n' );
            auto row = operator[]( i );
            for( int j = 0; j < columns.Size(); ++j )
            {
                if( j ) output.Append( ",\t" );
                row[ j ].Dump( output );
            }
        }
    }


    DbRow DbTable::operator[]( int rowIdx )
    {
#if __DEBUG
        if( rowIdx > rowCount )
        {
            throw std::exception( "out of range: rowIdx" );
        }
#endif
        return DbRow{ this, rowIdx };
    }
    DbRow DbTable::operator[]( int rowIdx ) const
    {
#if __DEBUG
        if( rowIdx > rowCount )
        {
            throw std::exception( "out of range: rowIdx" );
        }
#endif
        return DbRow{ const_cast<DbTable*>( this ), rowIdx };
    }

    int DbTable::GetIndexByColumnName( String const& name )
    {
        for( int i = 0; i < columns.Size(); ++i )
        {
            if( columns[ i ].name == name ) return i;       // todo: hash 优化？
        }
        return -1;
    }

}
